Create the TransactionsProof request

viquezclaudio opened the pull request:
Create a new request, in order to handle Transaction inclusion proofs.

This is necessary for web-clients to request transaction inclusion proofs from history nodes.

viquezclaudio committed:
Create the TransactionsProof request

Create a new request, in order to handle Transaction inclusion proofs.

sisou commented:
Will there also be a method on the consensus to request the proof of a specific transaction? Or does that come later in a separate PR after this one merges the message type only?

viquezclaudio commented:
That will be part of this PR too, so that the requests can be performed from JS

viquezclaudio committed:
Move extended transactions and inherents to primitives::transaction

- Inherents are a special type of transaction, so they should belong to the transaction crate
- Extended transaction is a wrapper around regular transactions and inherents
- This changes are necessary in order to expose and access extended transactions from a web client

jsdanielh commented on blockchain/src/history/history_store.rs:
```diff
@@ -17,11 +16,12 @@ use nimiq_mmr::mmr::proof::RangeProof;
 use nimiq_mmr::mmr::MerkleMountainRange;
 use nimiq_mmr::store::memory::MemoryStore;
 use nimiq_primitives::policy::Policy;
+use nimiq_transaction::extended_transaction::{ExtTxData, ExtendedTransaction};
+use nimiq_transaction::inherent::InherentType;
```
```suggestion
use nimiq_transaction::{extended_transaction::{ExtTxData, ExtendedTransaction}, inherent::InherentType};
```

jsdanielh commented on blockchain/src/history/history_store.rs:
```diff
@@ -17,11 +16,12 @@ use nimiq_mmr::mmr::proof::RangeProof;
 use crate::history::ordered_hash::OrderedHash;
-use crate::history::{ExtendedTransaction, HistoryTreeChunk, HistoryTreeProof};
-use crate::ExtTxData;
+use crate::history::{HistoryTreeChunk, HistoryTreeProof};
```
I'd collapse all of the `crate::history` uses in a single statement

jsdanielh commented on consensus/src/messages/handlers.rs:
```diff
@@ -300,3 +300,64 @@ impl Handle<ResponseChunk, Arc<RwLock<Blockchain>>> for RequestChunk {
+
+        let block = match blockchain
+            .chain_store
+            .get_block_at(block_height, false, None)
```
Sounds like we could use:
```
.ok().and_then(|block| { if block.is_macro() {
  // We expect a macro block
    Some(block)
  } else {
    None
}}
```

jsdanielh commented on consensus/src/messages/handlers.rs:
```diff
@@ -300,3 +300,64 @@ impl Handle<ResponseChunk, Arc<RwLock<Blockchain>>> for RequestChunk {
+
+        for hash in tx_hashes {
+            // Get all the extended transactions that correspond to this hash.
+            let mut extended_tx_vec = blockchain.history_store.get_ext_tx_by_hash(&hash, None);
```
Why is this `mut`?

jsdanielh replied:
Couldn't we have done something like:
`transactions.extend(blockchain.history_store.get_ext_tx_by_hash(&hash, None))`
?

jsdanielh commented on consensus/src/messages/mod.rs:
```diff
@@ -3,9 +3,11 @@ use std::fmt::{Debug, Formatter};
-use nimiq_blockchain::HistoryTreeChunk;
+use nimiq_blockchain::{HistoryTreeChunk, HistoryTreeProof};
 use nimiq_hash::Blake2bHash;
+use nimiq_keys::Address;
```
```suggestion
#[cfg(feature = "full")]
use nimiq_keys::Address;
```

jsdanielh commented on consensus/src/messages/mod.rs:
```diff
@@ -3,9 +3,11 @@ use std::fmt::{Debug, Formatter};
 use nimiq_hash::Blake2bHash;
+use nimiq_keys::Address;
 use nimiq_network_interface::request::{RequestCommon, RequestMarker};
+use nimiq_transaction::extended_transaction::ExtendedTransaction;
```
```suggestion
#[cfg(feature = "full")]
use nimiq_transaction::extended_transaction::ExtendedTransaction;
```

jsdanielh commented on consensus/src/messages/mod.rs:
```diff
@@ -202,3 +208,48 @@ impl RequestCommon for RequestHead {
+#[cfg(feature = "full")]
+impl RequestCommon for RequestTransactionsProof {
+    type Kind = RequestMarker;
+    const TYPE_ID: u16 = 212;
```
212 is already occupied by `RequestChunk`

jsdanielh commented on consensus/src/sync/history/cluster.rs:
```diff
@@ -8,11 +8,12 @@ use std::task::{Context, Poll};
 use futures::{FutureExt, Stream, StreamExt};
 use lazy_static::lazy_static;
 use nimiq_blockchain_interface::{AbstractBlockchain, PushError, PushResult};
+use nimiq_transaction::extended_transaction::ExtendedTransaction;
```
I'd place `nimiq` uses in the next section

jsdanielh commented on primitives/account/src/staking_contract/traits.rs:
```diff
@@ -8,6 +8,7 @@ use nimiq_primitives::{account::AccountError, coin::Coin, policy::Policy, slots:
 use nimiq_transaction::account::staking_contract::{
     IncomingStakingTransactionData, OutgoingStakingTransactionProof,
 };
+use nimiq_transaction::inherent::{Inherent, InherentType};
 use nimiq_transaction::Transaction;
```
```suggestion
use nimiq_transaction::{account::staking_contract::{
    IncomingStakingTransactionData, OutgoingStakingTransactionProof,
}, inherent::{Inherent, InherentType}, Transaction};
```

jsdanielh commented on primitives/account/tests/accounts.rs:
```diff
@@ -1,13 +1,11 @@
 use std::convert::TryFrom;
 use std::time::Instant;
-
-use rand::{rngs::StdRng, SeedableRng};
```
This was actually ok, we list `std` imports apart

jsdanielh commented on primitives/transaction/Cargo.toml:
```diff
@@ -20,12 +20,16 @@ strum_macros = "0.24"
 thiserror = "1.0"
 
 beserial = { path = "../../beserial", features = ["derive"] }
+
```
This line is unnecessary: `beserial` is an internal crate as well

jsdanielh commented on primitives/transaction/Cargo.toml:
```diff
@@ -20,12 +20,16 @@ strum_macros = "0.24"
+nimiq-mmr = { path = "../mmr" }
 nimiq-network-interface = { path = "../../network-interface" }
 nimiq-primitives = { path = "..", features = ["account", "coin", "networks", "policy", "serde-derive"] }
+
```
This line is not needed as well

jsdanielh commented on primitives/transaction/src/extended_transaction.rs:
```diff
@@ -3,15 +3,17 @@ use std::fmt;
 
 use beserial::{Deserialize, ReadBytesExt, Serialize, SerializingError, WriteBytesExt};
-use nimiq_account::{Inherent, InherentType};
+
```
This line is unnecessary: `beserial` is an internal crate

jsdanielh commented on primitives/transaction/src/inherent.rs:
```diff
@@ -4,10 +4,13 @@ use beserial::{Deserialize, ReadBytesExt, Serialize, SerializingError, WriteByte
 #[derive(Clone, Debug, Eq, PartialEq, Copy, Serialize, Deserialize)]
 #[repr(u8)]
+/// Enum that represents the different types of inherents.
+/// They moment when those inherents are applied depends upon the specific type of inherent.
```
```suggestion
/// The moment when these inherents are applied depends upon the specific type of inherent.
```

jsdanielh replied:
I'd add member `rustdoc` as well

jsdanielh commented on primitives/transaction/src/inherent.rs:
```diff
@@ -28,6 +31,9 @@ impl InherentType {
 #[derive(Clone, Debug, Eq, PartialEq)]
+/// An inherent is a special kind of transaction
+/// It does not have a sender account, but it does have a target address and a value that is transfered
+/// For instance, a reward (for producing blocks) is an inherent
```
I'd add member `rustdoc` as well
